src/unit.h

```diff
@@ -6,25 +6,162 @@
+
+/// unit constants
+BOOST_UNITS_STATIC_CONSTANT(angstrom, wavelength);
+BOOST_UNITS_STATIC_CONSTANT(angstroms, wavelength);
+BOOST_UNITS_STATIC_CONSTANT(meV, energy);
+BOOST_UNITS_STATIC_CONSTANT(meVs, energy);
+BOOST_UNITS_STATIC_CONSTANT(microsecond, tof);
+BOOST_UNITS_STATIC_CONSTANT(microseconds, tof);
+
+} // namespace tof
+} // namespace neutron
+
+// Define conversion factors: the conversion will work both ways with a single
+// macro
+
+// Convert angstroms to SI meters
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(neutron::tof::wavelength_base_unit,
+                                     boost::units::si::length, double, 1.0e-10);
+// Convert meV to SI Joule
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(
+    neutron::tof::energy_base_unit, boost::units::si::energy, double,
+    1.0e-3 * boost::units::si::constants::codata::e.value().value());
+// Convert tof microseconds to SI seconds
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(neutron::tof::tof_base_unit,
+                                     boost::units::si::time, double, 1.0e-6);
+
+template <>
+struct boost::units::base_unit_info<neutron::tof::wavelength_base_unit> {
+  static std::string name() { return "angstroms"; }
+  static std::string symbol() { return "AA"; }
+};
+
+template <>
+struct boost::units::base_unit_info<neutron::tof::counts_base_unit> {
+  static std::string name() { return "counts"; }
+};
+
+template <>
+struct boost::units::base_unit_info<neutron::tof::energy_base_unit> {
+  static std::string name() { return "milli-electronvolt"; }
+  static std::string symbol() { return "meV"; }
+};
+
+template <> struct boost::units::base_unit_info<neutron::tof::tof_base_unit> {
+  static std::string name() { return "microseconds"; }
+  static std::string symbol() { return "us"; }
+};
+
+namespace {
+boost::units::si::dimensionless none;
+boost::units::si::length m;
+boost::units::si::area m2;
+boost::units::si::time s;
+boost::units::si::mass kg;
+neutron::tof::counts counts;
+neutron::tof::wavelength lambda;
+neutron::tof::energy mev;
+neutron::tof::tof tof;
+} // namespace
+
 class Unit {
 public:
+  // Define a std::variant which will hold the set of allowed units.
+  // Any unit that does not exist in the variant will either fail to compile or
+  // throw a std::runtime_error during operations such as multiplication or
+  // division.
+  //
+  // TODO: maybe it is possible to create a helper that will automatically
+  // generate the squares for variance?
+  // The following was attempted but did not succeed:
+  //  template <class... Ts> struct unit_and_variance {
+  //  using type =
+  //    std::variant<Ts..., decltype(std::declval<Ts>()*std::declval<Ts>())...>;
+  //  };
+  typedef std::variant<decltype(none),     // Dimensionless [ ]
+                       decltype(m),        // Length [m]
+                       decltype(m2),       // Area [m^2]
+                       decltype(s),        // Time [s]
+                       decltype(kg),       // Mass [kg]
```

Comment:
With the shortcut definitions, I'd recommend to remove most of the comments. I can see from the code that it is `kg`. :)